File: src/backend/modules/m6_groupware/schemas/approvals.py
"""
M6 그룹웨어 — 결재 Pydantic 스키마
"""
from typing import Optional
from pydantic import BaseModel, Field


# ── 결재선 템플릿 ──

class TemplateLineCreate(BaseModel):
    step_order: int = 1
    approver_id: str
    role_label: Optional[str] = None
    line_type: str = "approval"  # approval / reference


class TemplateLineResponse(BaseModel):
    id: str
    step_order: int
    approver_id: str
    approver_name: Optional[str] = None
    role_label: Optional[str] = None
    line_type: str


class TemplateCreate(BaseModel):
    name: str
    document_type: Optional[str] = None
    description: Optional[str] = None
    lines: list[TemplateLineCreate] = Field(min_length=1)


class TemplateResponse(BaseModel):
    id: str
    name: str
    document_type: Optional[str] = None
    description: Optional[str] = None
    is_active: bool
    lines: list[TemplateLineResponse] = []
    created_at: Optional[str] = None


# ── 결재 요청 ──

class ApprovalStepCreate(BaseModel):
    step_type: str = "approval"  # approval / reference
    step_order: int = 1
    approver_id: str
    role_label: Optional[str] = None


class ApprovalStepResponse(BaseModel):
    id: str
    step_type: str
    step_order: int
    approver_id: str
    approver_name: Optional[str] = None
    role_label: Optional[str] = None
    status: str
    comment: Optional[str] = None
    acted_at: Optional[str] = None


class ApprovalRequestCreate(BaseModel):
    title: str
    document_type: str = "general"
    content: Optional[dict] = None
    amount: Optional[float] = None
    reference_type: Optional[str] = None
    reference_id: Optional[str] = None
    steps: list[ApprovalStepCreate] = Field(min_length=1)


class ApprovalListItem(BaseModel):
    id: str
    request_no: str
    title: str
    document_type: str
    amount: Optional[float] = None
    status: str
    requester_id: str
    requester_name: Optional[str] = None
    reference_type: Optional[str] = None
    reference_id: Optional[str] = None
    current_step: Optional[int] = None
    total_steps: Optional[int] = None
    created_at: Optional[str] = None


class ApprovalResponse(ApprovalListItem):
    content: Optional[dict] = None
    steps: list[ApprovalStepResponse] = []


class ApprovalActionData(BaseModel):
    comment: Optional[str] = None

File: src/frontend/src/api/groupware/approvals.ts
/**
 * M6 그룹웨어 — 결재/템플릿 API 호출 함수
 */
import api from '../client';

/* ── 타입 정의 ── */

export interface TemplateLine {
  id?: string;
  step_order: number;
  approver_id: string;
  approver_name?: string;
  role_label?: string;
  line_type: string; // approval / reference
}

export interface Template {
  id: string;
  name: string;
  document_type?: string;
  description?: string;
  is_active: boolean;
  lines: TemplateLine[];
  created_at?: string;
}

export interface ApprovalStep {
  id: string;
  step_type: string;
  step_order: number;
  approver_id: string;
  approver_name?: string;
  role_label?: string;
  status: string;
  comment?: string;
  acted_at?: string;
}

export interface ApprovalListItem {
  id: string;
  request_no: string;
  title: string;
  document_type: string;
  amount?: number;
  status: string;
  requester_id: string;
  requester_name?: string;
  reference_type?: string;
  reference_id?: string;
  current_step?: number;
  total_steps?: number;
  created_at?: string;
}

export interface ApprovalDetail extends ApprovalListItem {
  content?: Record<string, unknown>;
  steps: ApprovalStep[];
}

/* ── 결재선 템플릿 API ── */

export async function listTemplates() {
  const res = await api.get('/groupware/templates');
  return res.data?.data ?? res.data;
}

export async function createTemplate(data: {
  name: string;
  document_type?: string;
  description?: string;
  lines: Omit<TemplateLine, 'id' | 'approver_name'>[];
}) {
  const res = await api.post('/groupware/templates', data);
  return res.data?.data ?? res.data;
}

export async function deleteTemplate(id: string) {
  const res = await api.delete(`/groupware/templates/${id}`);
  return res.data?.data ?? res.data;
}

/* ── 결재 요청 API ── */

export async function listApprovals(params?: {
  status?: string;
  document_type?: string;
  page?: number;
  size?: number;
}) {
  const res = await api.get('/groupware/approvals', { params });
  return res.data?.data ?? res.data;
}

export async function getApproval(id: string) {
  const res = await api.get(`/groupware/approvals/${id}`);
  return res.data?.data ?? res.data;
}

export async function createApproval(data: {
  title: string;
  document_type?: string;
  content?: Record<string, unknown>;
  amount?: number;
  reference_type?: string;
  reference_id?: string;
  steps: { step_type: string; step_order: number; approver_id: string; role_label?: string }[];
}) {
  const res = await api.post('/groupware/approvals', data);
  return res.data?.data ?? res.data;
}

export async function myRequests(params?: { page?: number; size?: number }) {
  const res = await api.get('/groupware/approvals/my-requests', { params });
  return res.data?.data ?? res.data;
}

export async function myApprovals(params?: { status?: string }) {
  const res = await api.get('/groupware/approvals/my-approvals', { params });
  return res.data?.data ?? res.data;
}

export async function myReferences() {
  const res = await api.get('/groupware/approvals/my-references');
  return res.data?.data ?? res.data;
}

export async function approveRequest(id: string, comment?: string) {
  const res = await api.patch(`/groupware/approvals/${id}/approve`, { comment });
  return res.data?.data ?? res.data;
}

export async function rejectRequest(id: string, comment?: string) {
  const res = await api.patch(`/groupware/approvals/${id}/reject`, { comment });
  return res.data?.data ?? res.data;
}
